### DISCUSSION TITLE
What is the easiest way to capture participant ID from a url and save to the .csv?

### DISCUSSION THREAD

**Discussion by Cluster-Bomb** [jspsych/jsPsych#737] [Q&A] [answered]
Hi there, 

Need to capture the participant ID from the experiment url and somehow save it to the .csv along with all the other trial data. 

I have no problem capturing the ID using  jsPsych.data.getURLVariable(var_name) - however I dont know how to write it to the .csv. 

Cheers

**Comment by jodeleeuw** [comment, marked as answer]
[jsPsych.data.addProperties()](https://www.jspsych.org/core_library/jspsych-data/#jspsychdataaddproperties) will do what you are looking for.

```js
var subject_id = jsPsych.data.getURLVariable('PROLIFIC_PID');
jsPsych.data.addProperties({
	subject_id: subject_id
});
```

**Reply by kupiqu** [reply]
Oops, that's right, I double-checked my script after your comment. 

What is actually saving the prolific information in my case is:

```js
        {
           "type": ["call-function"],
           "func": function() { jsPsych.data.write( jsPsych.data.urlVariables() ); },
           "data": {
             "stage": ["prolific_info"]
           }
         },
```

This also saves other info in addition to the participant, like session, study_id, and preview information.

However, not sure why, the call-function appears in two lines of the saved csv file, one with the proper information and another line appears empty. I have no clue where this second line comes from.

I will check if @jodeleeuw's method above fixes the issue.

**Reply by kupiqu** [reply]
I think I understand what's going on now:

- if I use `jsPsych.data.addProperties` instead of `jsPsych.data.write` above, it adds the prolific information to all lines in the csv file, and there's only one `prolific_info` line. 

- if I use the "write" version as it is above, there's two lines for `prolific_info` because there is one additional line that is written automatically at the end of the trial, which is empty because the properties have not been added anywhere.

I sort of like the second minimalistic option better so the prolific info is not spread all around but it just appears in its line. However, I'd like to prevent the appearance of the second empty `prolific_info` line.

@jodeleeuw, is there a way to disable the automatic `jsPsych.data.write` call at the end of a trial?

**Reply by jodeleeuw** [reply]
No, you can't disable the data.write call because jsPsych makes the assumption that every plugin call gets associated with one entry in the data array. But the return value of the function in call-function is stored in the data, so you could add it by returning it from the function.

One reason to have the prolific id on every line of the data file is it keeps the data [tidy](https://vita.had.co.nz/papers/tidy-data.pdf). This makes it easy to interface with data analysis tools like R's `tidyverse` that expect tidy data. But obviously if your analysis doesn't require that then you can save a little space in the data by having the ID appear once. I just wanted to mention this here for others who are viewing this thread and wondering which approach to use.

**Reply by kupiqu** [reply]
Thanks @jodeleeuw  for the clarification!

> But the return value of the function in call-function is stored in the data, so you could add it by returning it from the function.

Unfortunately the value returned by `jsPsych.data.urlVariables()` is `[object Object]`

> One reason to have the prolific id on every line of the data file is it keeps the data tidy.

I do use R (in fact I use jaysire to build the experiment) and I would appreciate this feature to simplify my analysis. However, in my case, I still need to do some parsing as I want my analysis scripts to deal with two types of participants: online participants via Pavlovia-Prolific; and "local" participants performing the experiment in the lab.

For the latter, I ask participants to manually enter their own participant id using the jspsych-survey-text plugin and so far I couldn't find a way (in jaysire) to set the content of `responses` in a new column named `participant_id`.

I guess this would require porting `jsPsych.data.addProperties` to jaysire.

EDIT: `insert_property` can be used for adding `jsPsych.data.addProperties` in jaysire, however the problem is how to introduce a jsPsych variable in it

**Reply by jodeleeuw** [reply]
I haven't got too much experience with `jaysire` yet, but if you can hook into the `on_finish` event for the text entry trial, you can do something like this:

```js
var trial = {
	type: 'survey-text',
	on_finish: function(data){
		var id = JSON.parse(data.responses).Q0;
		jsPsych.data.addProperties({subject_id: id})
	}
}
```

**Reply by kupiqu** [reply]
I adapted your response to my case as follows and worked!

```r
    on_finish = insert_javascript(
      "function(data){
        var id = JSON.parse(data.responses).participant_id;
        jsPsych.data.addProperties({participant_id: id})
      }"
    )
```

This is actually great for my analysis, and a very nice know-how, thanks!!

**Comment by Cluster-Bomb** [comment]
Thanks - works great!
